### ISSUE TITLE
AWS doc improvments

### ISSUE THREAD

**Issue by benmccann** [GoogleCloudPlatform/kubernetes#7036]
I was looking at playing around with a k8s cluster on AWS, but am left with some questions that aren't answered in the [getting started guide](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/getting-started-guides/aws.md). CC @justinsb the resident AWS expert :-)

* Can I create my cluster with some number of nodes other than 4?
  * Looks like the script is calling [cluster/kube-up.sh](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/kube-up.sh) which in turn calls [cluster/aws/util.sh](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/aws/util.sh). That script is doing a ton. Looks like maybe it's setting up a SaltStack server which in turn provisions the cluster?
* Document current limitations
  * I see some ELB work that's in progress. Does that I mean I can only use k8s on AWS for workloads with no public-facing portion until this is done?
  * There are some general k8s limitations. E.g. it looks like you can't expand the size of you cluster. Is there a general list of things that are still in progress that can be linked to?
  * Probably a bunch of stuff I'm not aware of...
* Do I need [flannel](https://github.com/coreos/flannel#flannel)? Or is that already being handled by these scripts?
  * I've come across some postings that describe how k8s expects a subnet per host, which is provided only on GCE
* Explain high availability / running in multiple AZs
  * If you want your service to be highly available, it's typically recommended that you run in multiple AZs. I believe the current k8s recommendation is that you run a cluster per AZ? Unless the scheduler has some sort of data center awareness and anti-affinity concept it can use to ensure your multiple of your service don't all end up in the same AZ...

**Comment by brendandburns** [comment]
Some Answers:
  * You can create a cluster with different sizes, either set the ```$NUM_MINIONS``` environment variable before cluster turn-up, or edit ```cluster/aws/config-default.sh```

  * Yes, that script does the bulk of the work to do the setup of the servers.  And yes, Salt is used for a bunch of the binary install/machine configuration

  * No, you can create public facing services, but it's not automatic.  Once the ELB stuff is done k8s will be able to create the load balancer for you automatically.  For now, it's required that you manually set up external access, either by using ```hostPort``` to map a pod onto the host VM's IP address, or setting up the ELB yourself.

   * Yeah, we're working on dynamically sized clusters, but its not there yet.

   * No, you don't need flannel.  The default AWS install uses AWS VPC advanced routes for its networking.  AWS VPC Advanced networking is also capable of subnet per host.

   * The current recommendation is one cluster per AZ.  There have been some discussions about expanding our spreading (aka anti-affinity) to support AZs, it currently only spreads across machines.  We'll probably get that done after 1.0.

Speaking of 1.0, I'd highly recommend checking out the [roadmap](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/roadmap.md) for the state of features under development in the near term.  You can also read our v1.0 milestone for concrete issues.

Best
--brendan

**Comment by benmccann** [comment]
Thanks @brendandburns !  That's super helpful. A lot of my confusion was not finding the rest of the docs since they weren't linked to from the website. Filed a new issue (https://github.com/GoogleCloudPlatform/kubernetes/issues/7034) for that as you already saw. The Admin Guide looks like it will answer a lot of my questions.

I also missed `config-default.sh`, which would have been helpful so that I'd know how to bring up a different number of machines or bring it up in a different AZ. I've sent a PR to add a reference to that to the docs: https://github.com/GoogleCloudPlatform/kubernetes/pull/7088

The aws docs have a reference to aws-coreos.md, which uses Flannel. I wonder the coreos setup does not need to use Flannel given that AWS VPC Advanced networking is capable of subnet per host?  (Btw, is the advanced networking thing you referred to what they call ["enhanced networking"](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/enhanced-networking.html)?)

**Comment by justinsb** [comment]
Looks like brendan answered most of your questions (thanks brendan!)  If there are any remaining, feel free to re-ask them!

I've mostly been working on a configuration for AWS that uses Ubuntu as the host OS, and it doesn't use flannel.  I think theoretically you could swap out Ubuntu for CoreOS there, but in practice the current CoreOS configuration is based around flannel AFAIK.  I expect we will all start filling in the configuration matrix later, but I think right now everyone is pretty focused on getting a few key configurations up and running.

It looks like you also had a question about AWS enhanced networking - that's separate (it's just a performance thing).  AWS/Ubuntu uses custom routing in VPC.  It has its limitations (e.g. we have to pre-assign fixed CIDRs to each minion) but it works well.

**Comment by benmccann** [comment]
Yeah, I think I got all my AWS-specific questions figured out. Thank you guys for the help!

**Comment by bkeroackdsc** [comment]
Can I add a few of my similar AWS-related questions here?

- I don't want to use the automatic provisioning scripts, because I'd like to bring up a Kubernetes cluster on CoreOS in an existing subnet (in an extant VPC). How would networking work in this case? Do I need flannel?

- If I don't need flannel, how would I set up the VPC routing that you alluded to above? Assuming your using static entries in the subnet routing table, how would I tell k8s about that?

Thanks for any help!

**Comment by justinsb** [comment]
@bkeroackdsc I'd recommend opening a new issue.  For CoreOS questions, I'd ask @bakins.

The scripts for AWS/Ubuntu don't really cater to reusing an existing VPC or subnet.  The networking requirements are fairly tricky, and it just isn't something I/we have had time to prioritize.  In theory you can tag the VPC so that the script reuses it, but you also have to get the networking right and that is much harder.  AWS/Ubuntu doesn't use flannel, and it might be that flannel makes this easier by relaxing the networking requirements.

AWS/CoreOS's configuration is totally separate AFAIK, it goes through CloudFormation.  You best course of action there is to open a new issue and tag @bakins I think.

**Comment by bakins** [comment]
@bkeroackdsc Setting up Kubernetes on CoreOS "manually" on AWS. Is not too hard.  Feel free to open a new issue or ping me off line.
